jupyterhub/orm.py

```diff
@@ -423,6 +444,58 @@ def find(cls, db, name):
         """
         return db.query(cls).filter(cls.name == name).first()
 
+    def _check_data_only_column(self, column_name: str) -> bool:
+        """Check if a column is not a ForeignKey or in a
+        relationship.
+
+        Args:
+            column_name (str): ame of the column
+
+        Returns:
+            bool: returns `True` if the column is data-only,
+            returns `False` otherwise.
+        """
+        if hasattr(self, column_name):
+            column = getattr(Service, column_name)
+            if hasattr(column, 'foreign_keys'):
+                return len(column.foreign_keys) == 0
+            else:
+                return False
+        else:
+            return False
+
+    def update_column(self, column_name: str, value: any) -> bool:
```

Comment:
I don't quite follow what the get/update_column bits are for. I don't really think we should have logic based on whether fields are foreign keys or not. Can you explain more about that? I think the `input=True` trait metadata on the Service might already serve the purpose here, but I'm not sure.

Reply:
I need to save service data coming from the API to the service table in order to recreate the service from the database in the next Hub startup. And the `input=True` trait metadata and the non-foreign key are not in sync. 
For example, the `oauth_client_id` is an `input` trait of the `Service` object:
```python
    oauth_client_id = Unicode(
        help="""OAuth client ID for this service.

        You shouldn't generally need to change this.
        Default: `service-<name>`
        """
    ).tag(input=True)
```
But in the `Service` ORM, it comes from an `OAuthClient` table so I don't need to save this column.
```python
    oauth_client_id = Column(
        Unicode(255),
        ForeignKey(
            'oauth_clients.identifier',
            ondelete='SET NULL',
        ),
    )
```
When I recreate the service from the service database at Hub startup, for the input trait that I can read from the table, I can set it directly to the service object, and for the remaining traits, I need to set it manually.

```python
        traits = service.traits(input=True)
        for key in traits:
            value = orm_service.get_column(key)
            if value is not None:
                setattr(service, key, value)

        if orm_service.oauth_client is not None:
            service.oauth_redirect_uri = orm_service.oauth_client.redirect_uri
            service.oauth_client_id = orm_service.oauth_client.identifier
            service.oauth_redirect_uri = orm_service.oauth_client.redirect_uri
```